packages/vscode-extension/src/debug/teamsAppInstallation.ts

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { VS_CODE_UI } from "../extension";
+import graphLoginInstance from "../commonlib/graphLogin";
+import * as commonUtils from "./commonUtils";
+import { ExtensionErrors, ExtensionSource } from "../error";
+import * as StringResources from "../resources/Strings.json";
+
+import axios from "axios";
+import { returnSystemError, returnUserError } from "@microsoft/teamsfx-api";
+
+export async function showInstallAppInTeamsMessage(detected: boolean): Promise<boolean> {
+  const message = `${
+    detected
+      ? StringResources.vsc.localDebug.installApp.detection
+      : StringResources.vsc.localDebug.installApp.description
+  }${StringResources.vsc.localDebug.installApp.guide}`;
+  const result = await VS_CODE_UI.showMessage(
+    "info",
+    message,
+    true,
+    StringResources.vsc.localDebug.installApp.installInTeams,
+    StringResources.vsc.localDebug.installApp.continue
+  );
+  if (result.isOk()) {
+    if (result.value === StringResources.vsc.localDebug.installApp.installInTeams) {
+      const loginStatus = await graphLoginInstance.getStatus();
+      if (loginStatus.accountInfo?.upn === undefined) {
+        throw returnUserError(
+          new Error("M365 account info not found"),
+          ExtensionSource,
+          ExtensionErrors.GetTeamsAppInstallationFailed
+        );
+      }
+      const debugConfig = await commonUtils.getDebugConfig(true);
+      if (debugConfig?.appId === undefined) {
+        throw returnUserError(
+          new Error("Debug config not found"),
+          ExtensionSource,
+          ExtensionErrors.GetTeamsAppInstallationFailed
+        );
+      }
+      const url = `https://teams.microsoft.com/l/app/${debugConfig?.appId}?installAppPackage=true&webjoin=true&login_hint=${loginStatus.accountInfo.upn}`;
+      await VS_CODE_UI.openUrl(url);
+      return await showInstallAppInTeamsMessage(false);
+    } else if (result.value === StringResources.vsc.localDebug.installApp.continue) {
+      const internalId = await getTeamsAppInternalId();
+      return internalId === undefined ? await showInstallAppInTeamsMessage(true) : true;
+    } else if (result.value === StringResources.vsc.localDebug.installApp.cancel) {
+      return false;
+    }
+  }
+  return false;
+}
+
+export async function getTeamsAppInternalId(): Promise<string | undefined> {
+  const loginStatus = await graphLoginInstance.getStatus();
```

Comment:
Since we added new permission scope `TeamsAppInstallation.ReadForUser`, do we need to update graph login?
@xiaolang124

Reply:
From my practice, adding this scope seems good enough.

Reply:
It's weird... In [code](https://github.com/OfficeDev/TeamsFx/blob/dev/packages/vscode-extension/src/commonlib/graphLogin.ts#L28) we only have `Application.ReadWrite.All` scope, so how could `TeamsAppInstallation.ReadForUser` work as well...

Reply:
One guess is that `Application.ReadWrite.All` includes `TeamsAppInstallation.ReadForUser` but have not found any official doc.
To be safety, should add `TeamsAppInstallation.ReadForUser` to our graph scopes

Reply:
What I mean is adding `TeamsAppInstallation.ReadForUser` to the graph scopes will work well. But not sure if there are some others works not considered.

Reply:
Oh I see... just not included in this PR, right?

Reply:
Yes, will add afterwards since this PR do not affect the current behavior.